apps/derive/elpi/param2.elpi

```diff
@@ -91,18 +99,27 @@ param-match (fun N T B) P1 PRM :- pi x\ not (B x = fun _ _ _), !,
   param T T1 TR, !,
   (pi x x1 xR\ param x x1 xR => param (B x) (B1 x1) (BR x x1 xR)), !,
   P1 = fun N T1 B1,
-  (pi z z1\ PRM z z1 = fun N T x\ fun N T1 x1\
-                       fun N {coq.subst-fun [x,x1] TR} xR\
+  derive.param2.names12R N N1 N2 NR,
+  (pi z z1\ PRM z z1 = fun N1 T x\ fun N2 T1 x1\
+                       fun NR {coq.subst-fun [x,x1] TR} xR\
 		       {coq.mk-app (BR x x1 xR) [z x, z1 x1]}).
 
 param-match (fun N T B) P1 PRM :-
   param T T1 TR, !,
   (pi x x1 xR\ param x x1 xR => param-match (B x) (B1 x1) (BR x x1 xR)), !,
   P1 = fun N T1 B1,
-  (pi z z1\ PRM z z1 = fun N T x\ fun N T1 x1\
-                       fun N {coq.subst-fun [x,x1] TR} xR\
+  derive.param2.names12R N N1 N2 NR,
+  (pi z z1\ PRM z z1 = fun N1 T x\ fun N2 T1 x1\
+                       fun NR {coq.subst-fun [x,x1] TR} xR\
 		       BR x x1 xR z z1).
 
+% helper to improve name hints
+pred push-names i:term, i:list (option id), o:term.
+push-names (prod N T P) [N'|LN] (prod N''' T P') :- !,
+  if (N = `_`, N' = some N'') (coq.id->name N'' N''') (N''' = N),
```

Comment:
Since
```
`x` = `y`
```
this line is equivalent to `if (true, N' = some N'') ...`.
Is this still ok? It seems so, but I want to be sure.

Reply:
maybe 
```
push-names (prod N T P) [some N'|LN] (prod N' T P') :- !, pi x\ push-names (P x) LN (P' x).
push-names (prod N T P) [none|LN] (prod N T P') :- !, pi x\ push-names (P x) LN (P' x).
push-names T _ T :- !.
```

Reply:
and in this case I would call it "rename"

Reply:
Damn, my goal was to test the "absence" of name N, but I guess you solution is fine (and much nicer anyway).

Reply:
you can try with `coq.name->id N "_"` but I'm not sure it works.
Anyway, if the Arguments command was given a name for an argument, I would just use it no matter what the type checker put there.

Reply:
Thanks, indeed makes sense, done that.